egs/librispeech/asr/simple_v1/decode.py

```diff
@@ -137,61 +139,91 @@ def main():
     lang_dir = Path('data/lang_nosp')
     symbol_table = k2.SymbolTable.from_file(lang_dir / 'words.txt')
     phone_symbol_table = k2.SymbolTable.from_file(lang_dir / 'phones.txt')
-    ctc_topo = build_ctc_topo(list(phone_symbol_table._id2sym.keys()))
-    ctc_topo = k2.arc_sort(ctc_topo)
+
+    phone_ids = get_phone_symbols(phone_symbol_table)
+    P = create_bigram_phone_lm(phone_ids)
+
+    phone_ids_with_blank = [0] + phone_ids
+    ctc_topo = build_ctc_topo(phone_ids_with_blank)
+
+    logging.debug("About to load model")
+    # Note: Use "export CUDA_VISIBLE_DEVICES=N" to setup device id to N
+    # device = torch.device('cuda', 1)
+    device = torch.device('cuda')
+    model = TdnnLstm1b(num_features=40,
+                       num_classes=len(phone_ids) + 1, # +1 for the blank symbol
+                       subsampling_factor=3)
+    model.P_scores = torch.nn.Parameter(P.scores.clone(), requires_grad=False)
+
+    checkpoint = os.path.join(exp_dir, 'epoch-9.pt')
+    load_checkpoint(checkpoint, model)
+    model.to(device)
+    model.eval()
+
+    P.scores = model.P_scores.cpu()
+    assert P.requires_grad is False
 
     if not os.path.exists(lang_dir / 'LG.pt'):
-        print("Loading L_disambig.fst.txt")
+        logging.debug("Loading L_disambig.fst.txt")
         with open(lang_dir / 'L_disambig.fst.txt') as f:
             L = k2.Fsa.from_openfst(f.read(), acceptor=False)
-        print("Loading G.fsa.txt")
+        logging.debug("Loading G.fsa.txt")
         with open(lang_dir / 'G.fsa.txt') as f:
             G = k2.Fsa.from_openfst(f.read(), acceptor=True)
         first_phone_disambig_id = find_first_disambig_symbol(phone_symbol_table)
         first_word_disambig_id = find_first_disambig_symbol(symbol_table)
         LG = compile_LG(L=L,
                         G=G,
-                        ctc_topo=ctc_topo,
                         labels_disambig_id_start=first_phone_disambig_id,
                         aux_labels_disambig_id_start=first_word_disambig_id)
         torch.save(LG.as_dict(), lang_dir / 'LG.pt')
     else:
-        print("Loading pre-compiled LG")
+        logging.debug("Loading pre-compiled LG")
         d = torch.load(lang_dir / 'LG.pt')
         LG = k2.Fsa.from_dict(d)
 
+    # the following graph operations are put outside of `compile_LG`
+    # since `P` contains parameters depending on models. P is changed
+    # whenever decode.py uses a model from a different epoch.
+    logging.debug("Building the denominator graph")
+    den = k2.intersect(ctc_topo, P).invert_()
+
+    logging.debug("Connecting the denominator graph")
+    den = k2.connect(den)
+    den = k2.arc_sort(den)
+
+    logging.debug("Composing")
+    LG = k2.compose(den, LG)
```

Comment:
We don't actually want to include P here, only ctc_topo.  The idea is that P is estimated maximum likelihood so behaves like a stochastic LM, and it should be OK to replace it with the "real" LM without too much disruption / mismatch.  By composing P with the "real" LM we are "counting twice".

Reply:
thanks, will remove it and compare the WERs.